### PR TITLE
Added pagination, sorting and filtering to the get all projects endpoint

### LINKED ISSUE #156: Pagination for the get all projects endpoint
**Is your feature request related to a problem? Please describe.**
The get all projects endpoint does not yet support pagination.

**Describe the solution you'd like**
Pagination for the get all projects endpoint

### LINKED ISSUE #157: Sorting for the get all projects endpoint
**Is your feature request related to a problem? Please describe.**
The get all projects endpoint does not yet support sorting.

**Describe the solution you'd like**
Sorting for the get all projects endpoint

### LINKED ISSUE #161: Filtering for the get all projects endpoint
**Is your feature request related to a problem? Please describe.**
The get all projects endpoint does not yet support filtering.

**Describe the solution you'd like**
Filtering for the get all projects endpoint

### PR DESCRIPTION
**Pull request by StijnGroenen**
## Description

Added pagination, sorting and filtering to the get all projects endpoint.

## Type of change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] This change requires a documentation update

## Checklist

-   [x] My code follows the style guidelines of this project
-   [ ] I did not update API Controllers, if I did, I added/changed Postman tests
-   [x] I have performed a self-review of my own code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have made corresponding changes to the documentation
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes
-   [x] I updated the changelog with an end-user readable description
-   [x] I assigned this pull request to the correct project board to update the sprint board

## Steps to Test or Reproduce

1. Add a query parameter to the get all projects endpoint (page, amountonpage, sortby, sortdirection, highlighted)
2. The response should comply with the added parameters

## Link to issue

Closes #156 
Closes #157 
Closes #161

### CODE REVIEW

**Comment by StijnGroenen** [comment]
Because we agreed on using CRLF line endings, my IDE automatically converts the line endings to CRLF if they aren't already. This is expected and correct behaviour.
The entire code base should have been converted to CRLF after we decided that we were going to use this type on line ending. But apparently the code base still contains files using the LF line endings, which is causing conflicts now.
I think this is because the default behaviour for Windows developers is that the code uses CRLF locally but commits LF line endings.
image

File: Models/Params.cs [hunk 1]
```diff
@@ -1,28 +1,28 @@
...
+* You can find a copy of the GNU Lesser General Public License
 * along with this program, in the LICENSE.md file in the root project directory.
 * If not, see https://www.gnu.org/licenses/lgpl-3.0.txt
 */
 
 namespace API.Resources
 {
 
-    public class SearchParams
+    public class Params
```
Comment (rubenb994) [inline Models/Params.cs:21]: This doesnt seem like a proper name. The name params does not make it clear to me what it is used for

File: API/Controllers/SearchController.cs [hunk 2]
```diff
@@ -1,119 +1,119 @@
...
+
+        private readonly IMapper mapper;
+
+        private readonly ISearchService searchService;
+
+        /// <summary>
+        ///     Initialize a new instance of SearchController
+        /// </summary>
+        /// <param name="searchService"></param>
+        /// <param name="mapper"></param>
```
Comment (rubenb994) [inline API/Controllers/SearchController.cs:46]: Why the usage of empty params documentation?

File: API/Controllers/SearchController.cs [hunk 3]
```diff
@@ -1,119 +1,119 @@
...
+        /// <returns>Search results</returns>
+        [HttpGet("internal/{query}")]
+        public async Task<IActionResult> SearchInternalProjects(string query,
+                                                                [FromQuery] RequestParamsResource parameters)
+        {
+            ProblemDetails problem = new ProblemDetails
+            {
+                Title = "Invalid search request."
+            };
+            if(query.Length < 0)
```
Comment (rubenb994) [inline API/Controllers/SearchController.cs:67]: A string can be null causing null pointer exceptions when calling the length

File: Services/Services/ProjectService.cs [hunk 4]
```diff
@@ -1,55 +1,118 @@
...
+    }
+
+    public class ProjectService : Service<Project>, IProjectService
+    {
+
+        public ProjectService(IProjectRepository repository) : base(repository) { }
+
+        protected new IProjectRepository Repository => (IProjectRepository) base.Repository;
+
+        /// <summary>
```
Comment (rubenb994) [inline Services/Services/ProjectService.cs:50]: IMO the documentation should be located at the interface and not the implementation

File: API/Controllers/SearchController.cs (hunk 3 above)
Comment (wotwot563) [reply API/Controllers/SearchController.cs:67]:
```suggestion
            if(string.IsNullOrEmpty(query))

```

File: API/Resources/RequestParamsResource.cs [hunk 5]
```diff
@@ -21,9 +21,9 @@ namespace API.Resources
 {
 
     /// <summary>
-    ///     Resource to describe parameters for Search Request
+    ///     Resource to describe parameters for projects request
```
Comment (wotwot563) [inline API/Resources/RequestParamsResource.cs:24]: unclear documentation

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,40 +1,43 @@
...
+- Added example unittests for other contributors [#74](https://github.com/DigitalExcellence/dex-backend/issues/74)
+- You can now authenticate swagger with the Identity server - [#101](https://github.com/DigitalExcellence/dex-backend/issues/101)
+- Added issue & pull request templates (bug & report) - [#11](https://github.com/DigitalExcellence/dex-backend/issues/11)
+- Changed errors to be compliant with RFC 7807, with guids for easy error searching - [#80](https://github.com/DigitalExcellence/dex-backend/issues/80)
+- Added highlighted filter to search endpoint - [#57](https://github.com/DigitalExcellence/dex-backend/issues/57)
+- Setup basic unit test framework to ensure that the core functionality of the application works [#65](https://github.com/DigitalExcellence/dex-backend/issues/65)
+- Added roles and authorization validation. - [#107](https://github.com/DigitalExcellence/dex-backend/issues/107)
+- Added unittests for UserRepository - [#121] (https://github.com/DigitalExcellence/dex-backend/issues/121)
+- Added unittests for UserService - [#121] (https://github.com/DigitalExcellence/dex-backend/issues/121)
```
Comment (wotwot563) [inline CHANGELOG.md:22]: broken links

File: Services/Services/ProjectService.cs [hunk 7]
```diff
@@ -1,55 +1,118 @@
...
+
+        Task<Project> FindWithUserAndCollaboratorsAsync(int id);
+
+        Task<int> ProjectsCount(Params searchParams);
+
+        Task<int> GetProjectsTotalPages(Params searchParams);
+
+    }
+
+    public class ProjectService : Service<Project>, IProjectService
```
Comment (wotwot563) [inline Services/Services/ProjectService.cs:43]: no documentation

**Comment by StijnGroenen** [comment]
@rubenb994 and @wotwot563, most of the comments are about parts that didn't actually change in this PR. These are marked as additions because of the line endings issue (#163). I will wait for a solution to that issue before I will make the changes to this PR.

**Comment by Brend-Smits** [comment]
> @rubenb994 and @wotwot563, most of the comments are about parts that didn't actually change in this PR. These are marked as additions because of the line endings issue (#163). I will wait for a solution to that issue before I will make the changes to this PR.

I don't think you should wait for a solution to that issue, there are people waiting for this PR and it's blocking frontend. 
When we have a solution for the issue regarding line endings, we can make a new branch and push changes to fix them across the repositories.

**Comment by StijnGroenen** [comment]
> > @rubenb994 and @wotwot563, most of the comments are about parts that didn't actually change in this PR. These are marked as additions because of the line endings issue (#163). I will wait for a solution to that issue before I will make the changes to this PR.
> 
> I don't think you should wait for a solution to that issue, there are people waiting for this PR and it's blocking frontend.
> When we have a solution for the issue regarding line endings, we can make a new branch and push changes to fix them across the repositories.

I can do that as well.
Would you like to merge it with the CRLF line endings (and therefore changing the entire file) or with the LF line ending (which is not allowed by the coding guidelines)?

**Comment by StijnGroenen** [comment]
> > > @rubenb994 and @wotwot563, most of the comments are about parts that didn't actually change in this PR. These are marked as additions because of the line endings issue (#163). I will wait for a solution to that issue before I will make the changes to this PR.
> > 
> > 
> > I don't think you should wait for a solution to that issue, there are people waiting for this PR and it's blocking frontend.
> > When we have a solution for the issue regarding line endings, we can make a new branch and push changes to fix them across the repositories.
> 
> I can do that as well.
> Would you like to merge it with the CRLF line endings (and therefore changing the entire file) or with the LF line ending (which is not allowed by the coding guidelines)?

I will temporarily change it back to the LF line endings. We can change them (and everything else) if needed after we have decided on that issue.

**Comment by StijnGroenen** [comment]
@rubenb994 and @wotwot563, I changed the line endings back to LF. The changes of this PR should be more clear now. Could you review it again?

File: API/Controllers/ProjectController.cs [hunk 8]
```diff
@@ -56,23 +57,56 @@ public ProjectController(IProjectService projectService, IUserService userServic
         /// <summary>
         ///     Get all projects.
         /// </summary>
+        /// <param name="parameters">The parameters to narrow down the projects</param>
         /// <returns></returns>
         [HttpGet]
-        public async Task<IActionResult> GetAllProjects()
+        public async Task<IActionResult> GetAllProjects([FromQuery] RequestParamsResource parameters)
```
Comment (wotwot563) [inline API/Controllers/ProjectController.cs:63]: parameter name is not clear

File: API/Controllers/SearchController.cs [hunk 9]
```diff
@@ -58,7 +58,7 @@ public SearchController(ISearchService searchService, IMapper mapper)
         /// <returns>Search results</returns>
         [HttpGet("internal/{query}")]
         public async Task<IActionResult> SearchInternalProjects(string query,
-                                                                [FromQuery] SearchRequestParamsResource parameters)
+                                                                [FromQuery] RequestParamsResource parameters)
```
Comment (wotwot563) [inline API/Controllers/SearchController.cs:61]: parameter name not clear

File: Services/Services/ProjectService.cs [hunk 10]
```diff
@@ -27,10 +30,14 @@ namespace Services.Services
...
 
-        Task<List<Project>> GetAllWithUsersAsync();
+        Task<List<Project>> GetAllWithUsersAsync(Params parameters);
 
         Task<Project> FindWithUserAndCollaboratorsAsync(int id);
 
+        Task<int> ProjectsCount(Params searchParams);
+
+        Task<int> GetProjectsTotalPages(Params searchParams);
+
```
Comment (wotwot563) [inline Services/Services/ProjectService.cs:40]: no documentation

File: Services/Services/ProjectService.cs [hunk 11]
```diff
@@ -27,10 +30,14 @@ namespace Services.Services
...
     }
 
     public class ProjectService : Service<Project>, IProjectService
```
Comment (wotwot563) [inline Services/Services/ProjectService.cs:43]: no documentation

File: Services/Services/ProjectService.cs [hunk 12]
```diff
@@ -40,9 +47,65 @@ public class ProjectService : Service<Project>, IProjectService
 
         protected new IProjectRepository Repository => (IProjectRepository) base.Repository;
 
-        public Task<List<Project>> GetAllWithUsersAsync()
+        /// <summary>
+        /// Get a list of all the projects
+        /// </summary>
+        /// <param name="parameters">The parameters to narrow down the results</param>
+        /// <returns>A list of all the projects</returns>
+        public Task<List<Project>> GetAllWithUsersAsync(Params parameters)
```
Comment (wotwot563) [inline Services/Services/ProjectService.cs:55]: create tests that prove that all the different search features work

File: API/Controllers/ProjectController.cs (hunk 8 above)
Comment (StijnGroenen) [reply API/Controllers/ProjectController.cs:63]: Is the description in the documentation above clear? If it is, this should be fine. If it is not, I will change it.

**Pull request merged**